src/pqueue.c

```diff
@@ -0,0 +1,294 @@
+/*
+ * Collections-C
+ * Copyright (C) 2013-2014 Srđan Panić <srdja.panic@gmail.com>
+ *
+ * This file is part of Collections-C.
+ *
+ * Collections-C is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Collections-C is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Collections-C.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "array.h"
+#include "pqueue.h"
+
+struct pqueue_s {
+    Array *v;
+
+    /*
+     * Memory management function pointers
+     */
+    void *(*mem_alloc)  (size_t size);
+    void *(*mem_calloc) (size_t blocks, size_t size);
+    void  (*mem_free)   (void *block);
+
+    /*
+     * Comparator function pointer, for compairing the elements of PQueue
+     */
+    int   (*comparator) (const void *a, const void *b);
+};
+
+static int cc_default_comparator(const void *a, const void *b);
+static size_t cc_parent(size_t a);
+static void cc_heapify(PQueue *pqueue, size_t index);
+static size_t cc_left(size_t a);
+static size_t cc_right(size_t a);
+
+
+/*
+ * Initializes the fields of PQueueConf to default values
+ * @param[in, out] conf PQueueConf structure that is being initialized
+ */
+
+
+void pqueue_conf_init (PQueueConf *conf)
+{
+    array_conf_init(&conf->cfg);
+    conf->comparator = cc_default_comparator;
+}
+
+/**
+ * Creates a new empty pqueue and returns a status code.
+ *
+ * @param[out] out pointer to where the newly created PQueue is to be stored
+ *
+ * @return CC_OK if the creation was successful, or CC_ERR_ALLOC if the
+ * memory allocation for the new PQueue structure failed.
+ */
+
+enum cc_stat pqueue_new (PQueue **out)
```

Comment:
It might be better to not use a default comparator, but rather make the user to provide one during pqueue creation, like in the case of the [treetable](https://github.com/srdja/Collections-C/blob/master/src/treetable.c#L80). The reason for this is to make it harder for the user to do the wrong thing which would be forgetting to set the proper comparator.